include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1477,22 +1590,26 @@ struct __parallel_merge_submitter<__internal::__optional_kernel_name<_Name...>>
     auto
     operator()(_ExecutionPolicy&& __exec, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __rng3, _Compare __comp) const
     {
-        auto __n = __rng1.size();
+        using _Size = oneapi::dpl::__internal::__difference_t<_Range3>;
+
+        auto __n_1 = __rng1.size();
         auto __n_2 = __rng2.size();
 
-        assert(__n > 0 || __n_2 > 0);
+        assert(__n_1 > 0 || __n_2 > 0);
 
         _PRINT_INFO_IN_DEBUG_MODE(__exec);
 
-        const ::std::size_t __chunk = __exec.queue().get_device().is_cpu() ? 128 : 8;
-        const auto __max_n = ::std::max(__n, static_cast<decltype(__n)>(__n_2));
-        const ::std::size_t __steps = ((__max_n - 1) / __chunk) + 1;
+        const unsigned int __chunk = __exec.queue().get_device().is_cpu() ? 128 : 4;
+        const _Size __n = __n_1 + __n_2;
+        const _Size __steps = ((__n - 1) / __chunk) + 1;
 
         auto __event = __exec.queue().submit([&](sycl::handler& __cgh) {
             oneapi::dpl::__ranges::__require_access(__cgh, __rng1, __rng2, __rng3);
             __cgh.parallel_for<_Name...>(sycl::range</*dim=*/1>(__steps), [=](sycl::item</*dim=*/1> __item_id) {
-                __full_merge_kernel()(__item_id.get_linear_id() * __chunk, __rng1, decltype(__n)(0), __n, __rng2,
-                                      decltype(__n_2)(0), __n_2, __rng3, decltype(__n)(0), __comp, __chunk);
+
+                const unsigned int __i_elem = __item_id.get_linear_id() * __chunk;
+                const auto __start = __find_start_point(__rng1, __rng2, __i_elem, __n_1, __n_2);
```

Comment:
Could there be any benefit to loading a work group's data into SLM with a coalesced access pattern before passing it to this function? Since work items can't coalesce their accesses in `__find_start_point`, putting it into SLM first may be beneficial.

Reply:
If I'm understanding the algorithm correctly, this only works if ALL input data fits into SLM.  

Also, we don't have any guarantees on the data here (and how they will be merged), its possible that you may be loading all the input data to end up using only a part of it within the work group.  It would take some investigation to understand if this is of benefit for a "normal" use case, and you would need to come up with a strategy to handle the case where only part of the input sequence fits into SLM, and/or a backup where you don't use SLM.

Reply:
Probably, is to try to use SLM makes sense... But in that case we have got the additional n1+n2 reading and n1+n2 additional writing. Together with fact that is almost no computation - just fast binary search + m steps of merging, it is no obvious that we have got a performance gain....
BTW, I tried pre-caching  'chunk' elements form the first seq and 'chunk' elements form the second seq to private memory (registers), and got the performance slow down to 10%.

Reply:
I think to have any performance benefit with SLM, we will have to look at the GPU Merge Path variant of Merge Path which uses a tiling strategy. This can probably be a future optimization if it is difficult to implement since this implementation already provides significant performance benefit.